### PR TITLE
Feature/281 passphrase words

### PR DESCRIPTION
**Pull request by sadiabbasi**

### CODE REVIEW

File: test/controllers/user.test.ts [hunk 1]
```diff
@@ -32,4 +32,25 @@ describe('User class', () => {
...
     });
+    describe('generateKeywords()', () => {
+        let user: User;
+
+        beforeEach(() => {
+            user = new User(new JsKeyManager(), jsStorageFactory);
+        });
+
+        it('generates random passphrase words', async () => {
+            const generatedKeywords = await user.generateRandomPassphrase();
```
Comment (theblockstalk) [inline test/controllers/user.test.ts:43]: `generateRandomPassphrase()` and `suggestPassphraseWord()` are not asynchronous

File: test/controllers/user.test.ts [hunk 2]
```diff
@@ -32,4 +32,25 @@ describe('User class', () => {
...
+
+            expect(generatedKeywords).toHaveLength(6);
+        });
+
+        it('generates suggested passphrase words', async () => {
```
Comment (theblockstalk) [inline test/controllers/user.test.ts:48]: Some more tests, what happens if user sends empty string. What happens if user sends a word that is not in the list. What happens if a user sends some text that does not have any results?

File: test/controllers/user.test.ts [hunk 3]
```diff
@@ -32,4 +32,25 @@ describe('User class', () => {
             ).toThrowError();
         });
     });
+    describe('generateKeywords()', () => {
```
Comment (theblockstalk) [inline test/controllers/user.test.ts:35]: Your describe statement is about several functions, not just generateKeyword(). Either create a describe statement for all functions, or remove it and keep tests together with the other user tests

File: src/sdk/util/passphrase.ts [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+    return randomKeywords;
+}
+
+export function generateAutoSuggestions(inputString: string): string[] {
+    const wordsLinting = wordList.filter((word) => word.trim() !== '');
+
+    const matchingSuggestions: string[] = wordsLinting
+        .filter((word: string) => word.toLowerCase().includes(inputString.toLowerCase()))
+        .slice(0, 4);
```
[the remark this answers is not in the document]
Comment (sadiabbasi) [reply src/sdk/util/passphrase.ts:28]: so it means it returns 4 matches word not more than four.

File: test/controllers/user.test.ts (hunk 1 above)
Comment (sadiabbasi) [reply test/controllers/user.test.ts:43]: in user.ts it is asynchronous

File: src/sdk/topPassphrases.ts
Comment (theblockstalk) [inline src/sdk/topPassphrases.ts:1]: name of File is misleading. englistPassphaseWords.ts

Comment (theblockstalk) [reply src/sdk/topPassphrases.ts:1]: It would also be good if you added a comment with a link to the source of Where you got these

File: src/sdk/topPassphrases.ts [hunk 5]
```diff
@@ -0,0 +1,2051 @@
+export const topPasswords = [
```
Comment (theblockstalk) [inline src/sdk/topPassphrases.ts:1]: Change this variable name

File: src/sdk/generatePassphraseWordList.js [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+import { fileURLToPath } from 'url';
+import path from 'path';
+
+// Get the directory path of the current module
+const currentModulePath = fileURLToPath(import.meta.url);
+const currentDirectory = path.dirname(currentModulePath);
+
+// Construct the absolute path to the passphraseWords.txt file
+const filePath = path.join(currentDirectory, 'passphraseWords.txt');
+const TopPassphrasePath = path.join(currentDirectory, 'topPassphrases.ts');
```
Comment (theblockstalk) [inline src/sdk/generatePassphraseWordList.js:11]: You don't need to do all of this stuff with files. Just import the exported variable

Comment (theblockstalk) [reply src/sdk/generatePassphraseWordList.js:11]: This whole file can be deleted

Comment (theblockstalk) [reply src/sdk/generatePassphraseWordList.js:11]: see https://github.com/Tonomy-Foundation/Tonomy-ID-SDK/blob/master/src/sdk/util/passwords.ts#L2

File: src/sdk/passphraseWords.txt
Comment (theblockstalk) [inline src/sdk/passphraseWords.txt:1]: delete this file

File: src/sdk/util/passphrase.ts [hunk 7]
```diff
@@ -0,0 +1,39 @@
+import { topPasswords } from '../topPassphrases';
+import { randomBytes } from './crypto';
+
+export function generateRandomKeywords(): string[] {
+    const randomIndices: number[] = [];
+    const maxIndex = topPasswords.length;
+
+    while (randomIndices.length < 6) {
+        // Generate a random 4-byte number
+        const randomBytesArray = randomBytes(4);
```
Comment (theblockstalk) [inline src/sdk/util/passphrase.ts:10]: It would be cleaner if you created a new function `randomNumber(min: number, max: number)` in crypto.ts

File: src/sdk/util/passphrase.ts [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+            randomIndices.push(randomIndex);
+        }
+    }
+
+    const randomKeywords: string[] = randomIndices.map((index) => topPasswords[index]);
+
+    return randomKeywords;
+}
+
+export function generateAutoSuggestions(inputString: string): string[] {
```
Comment (theblockstalk) [inline src/sdk/util/passphrase.ts:23]: Add some TSDoc to explain the input and output (That will make it clear that this returns maximum 4 words)

File: src/sdk/util/passphrase.ts [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+    inputString = inputString.toLowerCase();
+    const matchingSuggestions: string[] = topPasswords
+        .filter((word: string) => word.toLowerCase().includes(inputString))
+        .slice(0, 4);
+
+    return matchingSuggestions;
+}
+
+// Convert a byte array to a number
+function byteArrayToNumber(byteArray: Uint8Array): number {
```
Comment (theblockstalk) [inline src/sdk/util/passphrase.ts:37]: As mentioned above, better to move something like this into our crypto.ts library

Comment (theblockstalk) [reply src/sdk/util/passphrase.ts:37]: ["separation of conserns"](https://en.wikipedia.org/wiki/Separation_of_concerns)

File: src/sdk/util/crypto.ts [hunk 10]
```diff
@@ -11,6 +11,20 @@ export function randomBytes(bytes: number): Uint8Array {
     return rb(new Uint8Array(bytes));
 }
 
+export function randomNumber(min: number, max: number): number {
```
Comment (theblockstalk) [inline src/sdk/util/crypto.ts:14]: `max: number` can be much larger than the 4 random bytes
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/MAX_SAFE_INTEGER

the number of bytes should be based on the value of (max-min)

File: src/sdk/util/passphrase.ts [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+
+/**
+ * Generates an array of random keywords.
+ * @returns An array of random keywords with a maximum of 6 words.
+ */
+export function generateRandomKeywords(): string[] {
+    const randomIndices: number[] = [];
+
+    while (randomIndices.length < 6) {
+        const randomIndex = randomNumber(1, 2047);
```
Comment (theblockstalk) [inline src/sdk/util/passphrase.ts:12]: this can never create an index of 0

File: src/sdk/util/crypto.ts [hunk 12]
```diff
@@ -11,6 +11,20 @@ export function randomBytes(bytes: number): Uint8Array {
...
+    if (min > max) {
+        throwError('Min value cannot be greater than max value');
+    }
+
+    const range = max - min + 1;
+    const randomBytesArray = randomBytes(4);
```
Comment (theblockstalk) [inline src/sdk/util/crypto.ts]: `const randomBytesArray = randomBytes( Math.floor(Math.log(range, 256))+1 )`

File: src/sdk/util/crypto.ts [hunk 13]
```diff
@@ -11,6 +11,20 @@ export function randomBytes(bytes: number): Uint8Array {
     return rb(new Uint8Array(bytes));
 }
 
+export function randomNumber(min: number, max: number): number {
+    if (min > max) {
+        throwError('Min value cannot be greater than max value');
+    }
+
+    const range = max - min + 1;
```
Comment (theblockstalk) [inline src/sdk/util/crypto.ts:29]: `range = max - min`
